### PR TITLE
add vox3D backend, add fracturing, improve ship assembly/loading and fix a batched rendering issue I forgot about

### PR DESCRIPTION
**Pull request by Tribulla**
## What this PR does:
- [x] Bug fix 
- [x] Feature
- [x] Code refactor / improvement
- [x] API addition / improvement
- [ ] Compat with another mod

**Explain what this PR is doing:**

adds the vox3D physics backend and fixes some issue in batched I have forgotten about. This now also adds all the fracturing stuff and improves ship assembly/loading.

Edit: now also fixes a config method issue introduced by tiger with the config method changes.
---

## Modloaders this PR affects:
- [x] Forge
- [x] Fabric

## Where this PR was tested:
- [x] In-dev singleplayer
- [x] In-dev dedicated server
- [x] Out of dev singleplayer
- [ ] GameTest framework

## Breaking Changes
- [ ] Yes (describe)
- [x] No

---

## Additional Notes
Anything else reviewers might need to know:
Performance concerns, edge cases, etc

---

## Declaration
By submitting this PR, I affirm:  
- Code/data compiles and loads  
- Tests _(if any)_ pass  
- This PR will not brick worlds

### CODE REVIEW

**Comment by Thettigerking** [comment]
NO WAY HE ACTUALLY PUSHED HIS CHANGES???

File: common/src/test/kotlin/org/valkyrienskies/mod/api/config/VSConfigApiBinaryCompatibilityTest.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    @Test
+    fun `legacy forge config builder signature remains available`() {
+        val classResource = "org/valkyrienskies/mod/api/config/VSConfigApi.class"
+        val classBytes = checkNotNull(javaClass.classLoader.getResourceAsStream(classResource)).use { it.readBytes() }
+        val classFile = String(classBytes, StandardCharsets.ISO_8859_1)
+        val legacyDescriptor = "(Lorg/valkyrienskies/core/internal/config/VsiConfigModelCategory;" +
+            "Lnet/minecraftforge/common/ForgeConfigSpec\$Builder;Lkotlin/jvm/functions/Function2;)" +
+            "Lnet/minecraftforge/common/ForgeConfigSpec\$Builder;"
+
+        assertTrue(classFile.contains(legacyDescriptor), "Missing Eureka-compatible buildForgeConfigSpec overload")
```
Comment (blockninja124) [inline common/src/test/kotlin/org/valkyrienskies/mod/api/config/VSConfigApiBinaryCompatibilityTest.kt]: why do we specify eureka in the error? other addons also used the old config system

Comment (Tribulla) [reply common/src/test/kotlin/org/valkyrienskies/mod/api/config/VSConfigApiBinaryCompatibilityTest.kt]: cus eureka was the first mod I realised was broken so when I was fixing it I wrote it before realising that all addons using the old system are broken

File: common/src/main/kotlin/org/valkyrienskies/mod/common/config/VSGameConfig.kt [hunk 2]
```diff
@@ -129,16 +129,16 @@ object VSGameConfig {
...
         var dynamicShipLighting = false
 
-        @ConfigEntry(description = "Project ships into the world's lighting at render time so ships occlude sunlight on the ground beneath them and ship-internal torches illuminate nearby world blocks (sodium/embeddium only). Experimental — overrides sodium's stock world-chunk shader. Disable for the default vanilla behavior where ships don't affect world lighting.")
+        @ConfigEntry(description = "Project ships into the world's lighting at render time so ships occlude sunlight on the ground beneath them and ship-internal torches illuminate nearby world blocks (sodium/embeddium only). Unlike the other ship-lighting options this affects the WORLD's chunks rather than ship blocks, so it applies to ALL ship renderers, the BATCHED default included. Experimental — overrides sodium's stock world-chunk shader. Disable for the default vanilla behavior where ships don't affect world lighting.")
```
Comment (blockninja124) [inline common/src/main/kotlin/org/valkyrienskies/mod/common/config/VSGameConfig.kt]: Get rid of the flipping em dash atleast I beg of you

File: common/src/main/kotlin/org/valkyrienskies/mod/common/command/VSCommands.kt [hunk 3]
```diff
@@ -54,6 +55,7 @@ object VSCommands {
         SchematicCommand.register(vs)
         ScaleCommand.register(vs)
         SplittingCommand.register(vs)
+        ShootCommand.register(vs)
```
Comment (blockninja124) [inline common/src/main/kotlin/org/valkyrienskies/mod/common/command/VSCommands.kt]: Isn't the shoot command redundant with my impulse command

File: common/src/main/kotlin/org/valkyrienskies/mod/common/command/commands/ShootCommand.kt [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                    Quaterniond(),
+                    Vector3d(1.0, 1.0, 1.0),
+                    centerOfShip
+                )
+            )
+        )
+
+        val id = BuiltInRegistries.BLOCK.getKey(block)
+        source.sendSuccess({
+            val base = "Fired ${size}x${size}x${size} $id ship #${ship.id} at ${"%.1f".format(speed)} m/s"
```
Comment (blockninja124) [inline common/src/main/kotlin/org/valkyrienskies/mod/common/command/commands/ShootCommand.kt]: Make this lang

File: common/src/main/kotlin/org/valkyrienskies/mod/common/command/commands/ShootCommand.kt [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            if (speed < IMPACT_GATE_MPS) {
+                Component.literal("$base  (note: impact fracture needs ~${IMPACT_GATE_MPS.toInt()}+ m/s to trigger)")
```
Comment (blockninja124) [inline common/src/main/kotlin/org/valkyrienskies/mod/common/command/commands/ShootCommand.kt]: Make this lang

File: common/src/main/kotlin/org/valkyrienskies/mod/common/command/commands/ShootCommand.kt [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        )
+    }
+
+    private fun shoot(ctx: CommandContext<CommandSourceStack>, size: Int, speed: Double, block: Block): Int {
+        val source = ctx.source
+        val level = source.level
+        val shooter = source.entity
+        if (shooter == null) {
+            source.sendFailure(
+                Component.literal("/vs shoot must be run by an entity (it fires from your eyes along your look direction).")
```
Comment (blockninja124) [inline common/src/main/kotlin/org/valkyrienskies/mod/common/command/commands/ShootCommand.kt]: Make this lang

File: common/src/main/kotlin/org/valkyrienskies/mod/api/config/VSConfigApi.kt [hunk 7]
```diff
@@ -39,6 +39,37 @@ object VSConfigApi {
         return builder
     }
 
+    @JvmStatic
+    @JvmName("buildForgeConfigSpec")
+    fun buildForgeConfigSpecCompat(
```
Comment (blockninja124) [inline common/src/main/kotlin/org/valkyrienskies/mod/api/config/VSConfigApi.kt]: Should this be `@Deprecated`?

File: common/src/main/java/org/valkyrienskies/mod/mixin/server/MixinMinecraftServer.java [hunk 8]
```diff
@@ -218,6 +218,7 @@ private void preTick(final CallbackInfo ci) {
             }
         }
         */
+        // endregion
```
Comment (blockninja124) [inline common/src/main/java/org/valkyrienskies/mod/mixin/server/MixinMinecraftServer.java]: Duplicated end of region

File: common/src/main/kotlin/org/valkyrienskies/mod/common/command/VSCommands.kt (hunk 3 above)
Comment (Tribulla) [reply common/src/main/kotlin/org/valkyrienskies/mod/common/command/VSCommands.kt]: shoot instantly spawns the projectile and shoots it in the direction the player looks at a set velocity measured in m/s so it is simpler to use

File: common/src/main/kotlin/org/valkyrienskies/mod/common/config/VSGameConfig.kt (hunk 2 above)
Comment (Tribulla) [reply common/src/main/kotlin/org/valkyrienskies/mod/common/config/VSGameConfig.kt]: done

File: common/src/main/java/org/valkyrienskies/mod/mixin/server/world/MixinServerLevel.java [hunk 9]
```diff
@@ -423,6 +439,8 @@ private void postTick(final BooleanSupplier shouldKeepTicking, final CallbackInf
         if (VSCoreConfig.SERVER.getSp().getEnableSplitting()) {
             ValkyrienSkiesMod.splitHandler.tick(ServerLevel.class.cast(this));
         }
+        ImpactFractureHandler.INSTANCE.tick(ServerLevel.class.cast(this));
```
Comment (blockninja124) [inline common/src/main/java/org/valkyrienskies/mod/mixin/server/world/MixinServerLevel.java]: These are not guarded on any config, impact and fracturing will be run even if the user has it disabled? Seems bad for performance

File: common/src/test/kotlin/org/valkyrienskies/mod/api/config/VSConfigApiBinaryCompatibilityTest.kt (hunk 1 above)
Comment (Tribulla) [reply common/src/test/kotlin/org/valkyrienskies/mod/api/config/VSConfigApiBinaryCompatibilityTest.kt]: fixed

File: common/src/main/java/org/valkyrienskies/mod/mixin/entity/MixinEntity.java [hunk 10]
```diff
@@ -241,22 +241,41 @@ private void preGetEyePositionServer(final CallbackInfoReturnable<Vec3> cir) {
...
+                    final Vector3dc renderPosition = shipTransform.getShipToWorld()
+                        .transformPosition(relativePosition, new Vector3d());
+                    cir.setReturnValue(shouldRenderAtSqrDistance(vs$distanceToSqr(renderPosition, d, e, f)));
                 }
             }
         }
     }
 
+    @Unique
+    private static double vs$distanceToSqr(final Vector3dc position, final double x, final double y,
```
Comment (blockninja124) [inline common/src/main/java/org/valkyrienskies/mod/mixin/entity/MixinEntity.java]: Why do we need our own distanceToSqr? Aren't there like a hundred api methods for this

File: common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java [hunk 11]
```diff
@@ -236,6 +243,69 @@ public void invalidateSection(long sectionPos) {
...
+                // the hard cap. The retainAll above still drops it when the ship leaves view.
+                e.setValue(v <= WINDOW_EXPIRED ? WINDOW_EXPIRED : v);
+            }
+        }
+    }
+
+    public boolean shipNeedsResettle(long shipId) {
+        shipSettleSeenThisTick.add(shipId);
+        if (!shipSettleTicksLeft.containsKey(shipId)) {
+            shipSettleTicksLeft.put(shipId, SETTLE_WINDOW_TICKS); // first sight: open window
```
Comment (blockninja124) [inline common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java]: what window is this referring to? and what "first sight"? first time the player looks at the ship?

File: common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java [hunk 12]
```diff
@@ -236,6 +243,69 @@ public void invalidateSection(long sectionPos) {
...
+        if (!shipSettleTicksLeft.isEmpty()) {
+            shipSettleTicksLeft.keySet().retainAll(shipSettleSeenThisTick);
+        }
+        shipSettleSeenThisTick.clear();
+        if (!shipSettleTicksLeft.isEmpty()) {
+            final ObjectIterator<Long2IntMap.Entry> it = shipSettleTicksLeft.long2IntEntrySet().iterator();
+            while (it.hasNext()) {
+                final Long2IntMap.Entry e = it.next();
+                final int v = e.getIntValue() - 1;
+                // Pin an elapsed window at WINDOW_EXPIRED instead of removing the entry.
```
Comment (blockninja124) [inline common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java]: maybe its just me but I can't make heads or tails of what this comment is trying to say

File: common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java [hunk 13]
```diff
@@ -76,6 +76,13 @@ public class VsShipLightStorage {
 
     private LongSet requestedThisFrame = new LongOpenHashSet();
 
+    private static final int SETTLE_WINDOW_TICKS = 100; // ~5s hard cap after first sight
```
Comment (blockninja124) [inline common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java]: Why is there a hard cap? explain what might cause the cap to be needed

File: common/src/main/java/org/valkyrienskies/mod/common/render/batched/ShipRenderObject.java [hunk 14]
```diff
@@ -31,6 +31,13 @@ public final class ShipRenderObject implements AutoCloseable {
     private long lastChunkPollGameTime = Long.MIN_VALUE;
     private boolean built = false;
 
+    private static final int RELIGHT_SETTLE_QUIET_TICKS = 3;
+    private static final long RELIGHT_SETTLE_MAX_DELAY_TICKS = 200L; // ~10s cap for slow mass-spawn propagation
```
Comment (blockninja124) [inline common/src/main/java/org/valkyrienskies/mod/common/render/batched/ShipRenderObject.java]: what the heck is "mass-spawn propagation", and why is this a Long

**Review by blockninja124 — CHANGES_REQUESTED** [review]
I have some more questions

File: common/src/main/kotlin/org/valkyrienskies/mod/common/command/VSCommands.kt (hunk 3 above)
Comment (blockninja124) [reply common/src/main/kotlin/org/valkyrienskies/mod/common/command/VSCommands.kt]: fair enough

**Comment by blockninja124** [comment]
Before this can be merged, we _need_ to fix the core version being unavailable from maven. We can't have public VS failing to import for everyone. I fixed circle ci for VS core but physics_api_krunch is still broken.

File: common/src/main/java/org/valkyrienskies/mod/mixin/server/world/MixinServerLevel.java (hunk 9 above)
Comment (Tribulla) [reply common/src/main/java/org/valkyrienskies/mod/mixin/server/world/MixinServerLevel.java]: that is a limit in vox3D and vox3D already has so much performance headroom that it doesnt really matter (I will look into it at some point and make a note about it tho)

File: common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java (hunk 13 above)
Comment (Tribulla) [reply common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java]: it can decide that it wants to settle forever which just causes weird flickering and stuff so I added a cap

File: common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java (hunk 12 above)
Comment (Tribulla) [reply common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java]: that is only known by the owerworked me from a month or two ago. ima figure it out and leave a better comment.

File: common/src/main/java/org/valkyrienskies/mod/mixin/server/world/MixinServerLevel.java (hunk 9 above)
Comment (blockninja124) [reply common/src/main/java/org/valkyrienskies/mod/mixin/server/world/MixinServerLevel.java]: but its also not blocked on running only when vox3d is active. Like, vox3d may be fast, but this will still be slowing down the game and especially noticeable on other engines

File: common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java (hunk 11 above)
Comment (Tribulla) [reply common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java]: first sight is when the ship gets loaded

File: common/src/main/kotlin/org/valkyrienskies/mod/common/command/commands/ShootCommand.kt (hunk 4 above)
Comment (Tribulla) [reply common/src/main/kotlin/org/valkyrienskies/mod/common/command/commands/ShootCommand.kt]: fixed

File: common/src/main/kotlin/org/valkyrienskies/mod/common/command/commands/ShootCommand.kt (hunk 5 above)
Comment (Tribulla) [reply common/src/main/kotlin/org/valkyrienskies/mod/common/command/commands/ShootCommand.kt]: fixed

File: common/src/main/kotlin/org/valkyrienskies/mod/common/command/commands/ShootCommand.kt (hunk 6 above)
Comment (Tribulla) [reply common/src/main/kotlin/org/valkyrienskies/mod/common/command/commands/ShootCommand.kt]: fixed

File: common/src/main/resources/assets/valkyrienskies/lang/en_us.json [hunk 15]
```diff
@@ -93,6 +93,10 @@
   "command.valkyrienskies.schematic.load.fail.missing_structure": "Missing structure file: %s",
   "command.valkyrienskies.schematic.load.fail.bad_joint_type": "Unsupported joint type: %s",
   "command.valkyrienskies.schematic.open_folder.server": "This command can't be used on a dedicated server",
+  "command.valkyrienskies.shoot.must_be_entity": "/vs shoot must be run by an entity (it fires from your eyes along your look direction).",
+  "command.valkyrienskies.shoot.unknown_block": "Unknown block: %s",
+  "command.valkyrienskies.shoot.success": "Fired %dx%dx%d %s ship #%s at %s m/s",
```
Comment (blockninja124) [inline common/src/main/resources/assets/valkyrienskies/lang/en_us.json]: Btw it doesn't actually matter but its worth knowing:
Minecraft doesn't properly parse string templates in lang. Everything is converted to a %s and then the .toString value is put in. So %d and %s makes no difference.

File: common/src/main/kotlin/org/valkyrienskies/mod/api/config/VSConfigApi.kt (hunk 7 above)
Comment (Tribulla) [reply common/src/main/kotlin/org/valkyrienskies/mod/api/config/VSConfigApi.kt]: fixed

File: common/src/main/java/org/valkyrienskies/mod/mixin/server/MixinMinecraftServer.java (hunk 8 above)
Comment (Tribulla) [reply common/src/main/java/org/valkyrienskies/mod/mixin/server/MixinMinecraftServer.java]: fixed

Comment (Tribulla) [reply common/src/main/java/org/valkyrienskies/mod/mixin/server/MixinMinecraftServer.java]: wait, I can just resolve conversation /:

File: common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java (hunk 13 above)
Comment (blockninja124) [reply common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java]: What does the "settling" refer to in terms of rendering? I'm just curious cause idk

**Comment by Tribulla** [comment]
I sure do love it when github lags and then decides that me clicking mark resolved a few too many times on the same thing meant that all are resolved /:

File: common/src/main/java/org/valkyrienskies/mod/mixin/server/world/MixinServerLevel.java (hunk 9 above)
Comment (Tribulla) [reply common/src/main/java/org/valkyrienskies/mod/mixin/server/world/MixinServerLevel.java]: oh wait. yeah, this is a issue

File: common/src/main/java/org/valkyrienskies/mod/mixin/entity/MixinEntity.java (hunk 10 above)
Comment (Tribulla) [reply common/src/main/java/org/valkyrienskies/mod/mixin/entity/MixinEntity.java]: not sure why I did this but I fixed it

File: common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java (hunk 13 above)
Comment (Tribulla) [reply common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java]: batched works by creating meshes of the ships that are then rendered in batches but due to how it works it means that lighting can become unstable (flickery or just broken) when ships are loaded and settling lets it figure itself out.

File: common/src/main/java/org/valkyrienskies/mod/mixin/server/world/MixinServerLevel.java (hunk 9 above)
Comment (Tribulla) [reply common/src/main/java/org/valkyrienskies/mod/mixin/server/world/MixinServerLevel.java]: fixed

File: common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java (hunk 11 above)
Comment (Tribulla) [reply common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java]: gonna close this

File: common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java (hunk 12 above)
Comment (Tribulla) [reply common/src/main/java/org/valkyrienskies/mod/compat/sodium/light/VsShipLightStorage.java]: owerworked 💀

File: common/src/main/resources/assets/valkyrienskies/lang/en_us.json (hunk 15 above)
Comment (Tribulla) [reply common/src/main/resources/assets/valkyrienskies/lang/en_us.json]: ah

File: common/src/main/java/org/valkyrienskies/mod/common/render/batched/ShipRenderObject.java (hunk 14 above)
Comment (Tribulla) [reply common/src/main/java/org/valkyrienskies/mod/common/render/batched/ShipRenderObject.java]: When ships or large structures are assembled, loaded, or spawned, Minecraft's lighting engine asynchronously propagates sky and block light across chunk sections.

When chunks change structurally, relightRemeshArmed is set to re-mesh the ship once the lighting engine has settled (hasLightWork() is false for RELIGHT_SETTLE_QUIET_TICKS = 3 consecutive ticks).
However, if the light engine is continuously kept busy (for instance, during heavy chunk loading or when many ships/chunks are spawning at once), quiet ticks might never be reached.
RELIGHT_SETTLE_MAX_DELAY_TICKS is a fallback deadline (~10 seconds / 200 ticks) ensuring a remesh is forced even if the light engine never reports an idle state.


Had to ask AI cus I did not remember what this was and the reason it is a long is cus `level.getGameTime()` returns a long.

**Review by blockninja124 — APPROVED** [review]
Good to merge but **not until core is fixed**
